src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/LengthBucketsFrozenDictionary.cs

```diff
@@ -121,32 +119,33 @@ internal sealed class LengthBucketsFrozenDictionary<TValue> : FrozenDictionary<s
         private protected override ref readonly TValue GetValueRefOrNullRefCore(string key)
         {
             // If the length doesn't have an associated bucket, the key isn't in the dictionary.
-            int length = key.Length - _minLength;
-            if (length >= 0)
+            int bucketIndex = (key.Length - _minLength) * MaxPerLength;
+            if (bucketIndex >= 0 && bucketIndex < _lengthBuckets.Length)
             {
-                // Get the bucket for this key's length.  If it's null, the key isn't in the dictionary.
-                KeyValuePair<string, int>[][] lengths = _lengthBuckets;
-                if ((uint)length < (uint)lengths.Length && lengths[length] is KeyValuePair<string, int>[] subset)
+                ReadOnlySpan<int> bucket = _lengthBuckets.AsSpan(bucketIndex, MaxPerLength);
+                string[] keys = _keys;
+                TValue[] values = _values;
+
+                foreach (int index in bucket)
                 {
-                    // Now iterate through every key in the bucket to see whether this is a match.
+                    // -1 is used to indicate a null
+                    if (index < 0)
+                    {
+                        break;
+                    }
+
                     if (_ignoreCase)
                     {
-                        foreach (KeyValuePair<string, int> kvp in subset)
+                        if (StringComparer.OrdinalIgnoreCase.Equals(key, keys[index]))
                         {
-                            if (StringComparer.OrdinalIgnoreCase.Equals(key, kvp.Key))
-                            {
-                                return ref _values[kvp.Value];
-                            }
+                            return ref values[index];
                         }
                     }
                     else
                     {
-                        foreach (KeyValuePair<string, int> kvp in subset)
+                        if (key == keys[index])
                         {
-                            if (key == kvp.Key)
-                            {
-                                return ref _values[kvp.Value];
-                            }
+                            return ref values[index];
                         }
```

Comment:
Did you actively decide against lifting out the _ignoreCase check and having two loops?